### PR TITLE
[SPARK-16992][PYSPARK] Virtualenv for Pylint and pep8 in lint-python

### PR DESCRIPTION
**Pull request by gsemet**
Use a virtualenv for isolation and easy installation of `pep8` and `pylint` by the `lint-python` linter script.

This involves 2 changes in this script:
- it creates a virtualenv if we are not in one. This way, we can use pip to install development environment's dependencies (which are different from the "prod" environment, for instance we don't want to install pep or pylint on prod)
- dependencies are described in `requirements.txt`. This will "freeze" the version of the tools used. We don't want the HEAD starting breaking because a new version of pylint has been posted on pypi which add new checks. sphinx version is also "frozen".

This basically reverts 85a50a6352b72c4619d010e29e3a76774dbc0c71

Might have been a solution to SPARK-9385 (see #7704)

Lot of new tests introduced by the newer version of pylint have been disabled. I propose to fix issues in various pull requests (obviously most 'import' order errors should be fixed by my other pull requests such as #14830 for documentation examples, which is part of the effort on code style described in #14567). Each subsequent pull request will fix one or more error and reenable the according pylint check.

List of new disabled tests:
- bad-super-call
- consider-iterating-dictionary
- consider-using-enumerate
- eval-used
- exec-used
- invalid-length-returned
- misplaced-comparison-constant
- raising-bad-type
- redefined-variable-type
- trailing-newlines
- trailing-whitespace
- ungrouped-imports
- unnecessary-pass
- unneeded-not
- wrong-import-order
- wrong-import-position
## How was this patch tested?

Manual execution of `lint-python` on my machine.

### CODE REVIEW

File: python/pylintrc [hunk 1]
```diff
@@ -84,7 +84,85 @@ enable=
...
 
-disable=invalid-name,missing-docstring,protected-access,unused-argument,no-member,unused-wildcard-import,redefined-builtin,too-many-arguments,unused-variable,too-few-public-methods,bad-continuation,duplicate-code,redefined-outer-name,too-many-ancestors,import-error,superfluous-parens,unused-import,line-too-long,no-name-in-module,unnecessary-lambda,import-self,no-self-use,unidiomatic-typecheck,fixme,too-many-locals,cyclic-import,too-many-branches,bare-except,wildcard-import,dangerous-default-value,broad-except,too-many-public-methods,deprecated-lambda,anomalous-backslash-in-string,too-many-lines,reimported,too-many-statements,bad-whitespace,unpacking-non-sequence,too-many-instance-attributes,abstract-method,old-style-class,global-statement,attribute-defined-outside-init,arguments-differ,undefined-all-variable,no-init,useless-else-on-loop,super-init-not-called,notimplemented-raised,too-many-return-statements,pointless-string-statement,global-variable-undefined,bad-classmethod-argument,too-many-format-args,parse-error,no-self-argument,pointless-statement,undefined-variable,undefined-loop-variable
+# Keep one item per line in the following block!
```
Comment (Stibbons) [inline python/pylintrc]: I have actually added a lot of exclusion tests here to make the code pass:
- bad-super-call
- consider-iterating-dictionary
- consider-using-enumerate
- eval-used
- exec-used
- invalid-length-returned
- misplaced-comparison-constant
- raising-bad-type
- redefined-variable-type
- trailing-newlines
- trailing-whitespace
- ungrouped-imports
- unnecessary-pass
- unneeded-not
- wrong-import-order
- wrong-import-position

File: dev/lint-python [hunk 2]
```diff
@@ -26,30 +67,26 @@ PYLINT_REPORT_PATH="$SPARK_ROOT_DIR/dev/pylint-report.txt"
...
+
+# It is a good practice to upgrade pip to the latest version, however, sometimes it might break
+# something. Let freeze this version here. We cannot put it into requirements.txt since the
+# execution of this requirement files might depends on the pip version.
+PIP_VERSION="8.1.2"
 
 cd "$SPARK_ROOT_DIR"
 
+echo "Jumping into virtualenv $VIRTUAL_ENV_DIR..."
+$VIRTUALENV_EXEC $VIRTUAL_ENV_DIR || exit 1
```
Comment (Stibbons) [inline dev/lint-python]: Note: here I leave the virtualenv of the current environment. tell me if which solution you prefere:
- jump into a custom venv? (this proposal)
- fail if in a virtualenv
- only warn the user if already in a virtualenv
- no warning at all, the user knows what he/she does

Comment (holdenk) [inline dev/lint-python]: I think if we _do_ jump to a custom virtual env for syntax checking we should at least restore the users existing virtual env after.

File: dev/lint-python [hunk 3]
```diff
@@ -17,6 +17,47 @@
 # limitations under the License.
 #
 
+VIRTUAL_ENV_DIR="build/venv"
+if [[ ! -z $VIRTUAL_ENV ]]; then
+    echo "We are in the following virtualenv: $VIRTUAL_ENV. Leaving it..."
+    # Cannot call "deactivate" from a bash script, copy/paste the code from virtualenv 'activate'
```
Comment (holdenk) [inline dev/lint-python]: I think we could maybe simplify this using the answer from http://stackoverflow.com/questions/29545624/how-to-deactivate-virtualenv-from-a-bash-script
(e.g. source or .) - what do you think?

**Comment by holdenk** [comment]
Thanks for working to reenable pylint (super important IMHO). Perhaps @davies could take a look at this approach?
I like the building up of venv seems to help set up for your other work improving pep8 support although it is maybe a bit of a bigger change than needed locally.

File: dev/lint-python (hunk 2 above)
Comment (Stibbons) [inline dev/lint-python]: What do you mean with "after"? For the Sphinx build? I think we can do both in this virtual env, this would allow to build without having to do sudo pip install Sphinx-build.
And we can control precisely the version of sphinx in this requirements.txt

In anyway, this virtualenv will be left automatically at the end of the execution of this script.

File: dev/requirements.txt [hunk 4]
```diff
@@ -1,3 +1,5 @@
 jira==1.0.3
 PyGithub==1.26.0
 Unidecode==0.04.19
+pep8==1.7.0
```
Comment (Stibbons) [inline dev/requirements.txt]: Need to add the sphinx version here

**Comment by holdenk** [comment]
Oh wait I've been a little confused by the PR title - pylint is already reenabled and was only temporary disabled awhile back - maybe less important but would make sense in the context of the other PRs you've put together. Could you maybe give this PR a clearer name like "Use virtualenv for pep8" rather than "Reenable Pylint"?

File: dev/lint-python (hunk 3 above)
Comment (Stibbons) [inline dev/lint-python]: I am open to suggestion. If we are already inside a venv we can expect the user knows what it does and don't desactivate it.

File: dev/lint-python (hunk 2 above)
Comment (holdenk) [inline dev/lint-python]: Using the virtualenv for the sphinx build sounds reasonable if we are already doing it for pep8. I was thinking that since we explicitly leave the current virtual env if the user is in one we should try and restore it after the script has been run - but if that already happens automatically thats great.

**Comment by Stibbons** [comment]
Hum I don't see how it was reenabled... Where is it called? And I had many errors to ignore once I have reenabled it on the execution of lint-python.

I'll update the title. At least, being inside a venv allow to precisely control the version of pep and pylint, this is kind of critical since errors changes from one version to another

File: dev/lint-python (hunk 2 above)
Comment (Stibbons) [inline dev/lint-python]: Actually I'm thinking it may be better not to leave virtualenv if we are inside one. So we don't have this copy paste of the code for desactivating it.
What do you think?

**Comment by holdenk** [comment]
I mean we already have the pep8 version locked since we download a specific version with curl rather than depending on the system version.

**Comment by Stibbons** [comment]
Yes but for pylint you have many dependencies to update as well (astroid,...). At least with a virtualenv, pip does it for us :)
And every time I see an hard coded external url I am afraid it might break soon or later

**Comment by holdenk** [comment]
I also think virtualenv or similar is probably the right way to go for PySpark developer dependencies in the long term :)

**Comment by Stibbons** [comment]
Great ! By the way, I am also working on virtualenv and wheel support for PySpark job deployment (see #14180)

**Comment by holdenk** [comment]
Yah I've been semi-following https://github.com/apache/spark/pull/14180 & https://github.com/apache/spark/pull/13599 but haven't had much time to look at them closely.  I think something like what you and @zjffdu are working on could make a real improvement to the PySpark experience for users with custom libraries.

**Comment by Stibbons** [comment]
Great !

**Comment by holdenk** [comment]
cc @davies ?

**Comment by Stibbons** [comment]
Hello, sorry to bother you, but if this patch gets merged, I can work on the pylint errors and submit new PR I had to add in the ignore list of pylint.

If I reenable most of them, here is an extract of errors I get:
```
************* Module pyspark.shell
W: 78, 8: Use of exec (exec-used)
************* Module pyspark.cloudpickle
R:767,16: Redefinition of v type from property to staticmethod (redefined-variable-type)
************* Module pyspark.tests
R:782,12: Redefinition of a type from pyspark.rdd.RDD to pyspark.rdd.PipelinedRDD (redefined-variable-type)
R:783,12: Redefinition of b type from pyspark.rdd.RDD to pyspark.rdd.PipelinedRDD (redefined-variable-type)
************* Module pyspark.rdd
R:1467,12: Redefinition of ser type from pyspark.serializers.AutoBatchedSerializer to pyspark.serializers.BatchedSerializer (redefined-variable-type)
************* Module pyspark.context
R:130,12: Redefinition of self.serializer type from pyspark.serializers.AutoBatchedSerializer to pyspark.serializers.BatchedSerializer (redefined-variable-type)
************* Module pyspark.worker
W:104,17: Use of eval (eval-used)
************* Module pyspark.shuffle
C:732,32: Consider iterating the dictionary directly instead of calling .keys() (consider-iterating-dictionary)
************* Module pyspark.ml.tests
R:1459, 8: Redefinition of a type from pyspark.ml.linalg.DenseVector to pyspark.ml.linalg.SparseVector (redefined-variable-type)
************* Module pyspark.ml.linalg
E:683, 4: __len__ does not return non-negative integer (invalid-length-returned)
R:1079, 8: Redefinition of colPtrs type from list to str (redefined-variable-type)
R:1078, 8: Redefinition of rowIndices type from list to str (redefined-variable-type)
************* Module pyspark.streaming.kafka
R:132,12: Redefinition of ser type from pyspark.serializers.PairDeserializer to pyspark.serializers.AutoBatchedSerializer (redefined-variable-type)
************* Module pyspark.streaming.tests
R:419, 8: Redefinition of initial type from list to pyspark.rdd.RDD (redefined-variable-type)
E:506,12: Bad first argument 'StreamingListener' given to super() (bad-super-call)
************* Module pyspark.streaming.context
C:298,12: Consider using enumerate instead of iterating with range and len (consider-using-enumerate)
```

These are not big errors, but it would be quite easy to fix them and have pylint runing at each pullrequest.

Thanks

**Comment by holdenk** [comment]
So even though this passes tests - its showing up in the PR dashboard as not for me. It might be easier to get reviewers to take a look if it was listed correctly, if you merge in the latest master the tests will re-run or we can explicitly ask jenkins to retest this.

**Comment by Stibbons** [comment]
What is the PR dashboard ?

I usually rebase this patch one or twice a week, I'll do it tomorrow

**Comment by holdenk** [comment]
oh actually it seems like the PR dashboard updated its self (it lives at https://spark-prs.appspot.com/ and is sometimes used by reviewers to decide which PRs to look at).

**Comment by Stibbons** [comment]
Pull Request rebased

**Comment by Stibbons** [comment]
Hello. Does anyone had time to review this patch? Is there anything I should do to help it being merged? This reenables Pylint checks using `lint-python` script and there are some errors that would need to be fixed thanks pylint.


Thanks

**Comment by holdenk** [comment]
Jenkins retest this please

**Comment by holdenk** [comment]
Maybe @davies, @srowen, or @yhuai  will have some thoughts as people who have worked on python linting.

**Comment by holdenk** [comment]
I think this could be a good change for allowing more developers to onboard with PySpark - is there any interest in the current PySpark/Build focused committers [ @davies @srowen @rxin ] in seeing this change (or something similar)?

**Comment by Stibbons** [comment]
I would love to have a bit more feedback on this matter but it does not seem to interest core developers, sadly :(
It's a bit disappointing, seeing how Python support on Spark is great, being able to deploy job as easy as with java (i.e., developers prepares the job + describes all the dependencies independently) would be so useful for Spark. For the moment, we have to ask our IT guys to install a given python module on each Executor when one needs a new one. We found a way using a nfs share but this is not convient. Automatic virtualenv creation with pip installing all dependencies described in a requirements.txt + support for wheels and python distribution package (sdist or bdist) is so useful and scalable. Each job can use different libraries and even the same library with different version, just like what can occur for java jars with the --packages arguments.
I think of starting maintaining a fork of Spark with latest version of the source code, I would call it "Python Friendly Spark".

**Comment by srowen** [comment]
Python really isn't my area, so I don't feel qualified to evaluate this. It sounds reasonable and has only build-time consequences, so I imagine it's low-risk. Improve style checking is good. I can merge this if there aren't objections.

**Comment by Stibbons** [comment]
sorry I wrote on the wrong pull request, I was talking about the virtualenv support for executor (#14180) :(
This one is indeed only to reenable pylint. Merging this would allow me to submit new patch with minor code style corrections, and get rid of this weird manual pep8 package download. virtualenv and pip have been made to make life of python developers easy, make deployment reproductible and cheap to maintain, I just would like to instill this to the Spark project. There are many reason why playing with the PYTHONPATH is wrong.

**Comment by holdenk** [comment]
Right I think the virtualenv support for PySpark in general is interesting - but its a large enough change we would need to get some substantail reviewer time dedicated to it (and I don't see that happening with the current somewhat bare bones PySpark committer team) but this is a bit more of a clear win and since it only has Spark developer tool implications - not end user implications, it's a bit easier to do new things.

How about we see if @davies has some time to look at this until Tuesday but if not - and if @srowen or one of the committers who has worked more on build related things feels comfortable - since its only a build change it might make sense to not be blocked too strongly on a Python specific reviewer. I'll take another look through the change tomorrow myself as well.

**Comment by holdenk** [comment]
Also @Stibbons I'm sorry the process has been so challenging and thanks for hanging in with it :)

**Comment by srowen** [comment]
Just so I'm doubly clear, this PR is ready to go, does something useful by itself? that is it's not just like a change that only makes sense with others? I'm OK to merge it if the Python folks think it's all a thumbs up.

**Comment by holdenk** [comment]
This PR does somehting useful by its self namely it uses virtual env / `requirements.txt` to fix the versions of `pylint` and `pep8` and reenables `pylint` checking of the code. There is a setup cost associated with installing the virtual env the first time we run this, but only for developers who run `./dev/pylint`.

Since this changes effectively adds back some linter rules its possible that mainline has changed since the last jenkins run and a merge to master performs poorly. Thinking about it some more, since we aren't under a rush for this, I'd like to double check this on Tuesday/Thursday on some of my coworkers machines that run OSX just to be safe because I know we've had some differences in the past with linter behaviour their (although this PR should remove those differences going forward its probably not a bad idea to double check).

**Comment by Stibbons** [comment]
Thanks both of you for your reviews.

It also fix the version of sphinx for documentation build during the `lint-python` My point of view is that for **application**, all versions of tools used to build should be frozen (in `requirements.txt`), to ensure full reproductibility. Upgrading any tool from version A to version B should be as much controlled as any other change in the source code, since it may have unwanted behavior. Pylint adds and remove checks from time to time and you don't want you code to stop building when pylint release a new version on Pypi, without a easy way to go back to a known state.

I am ready to submit new PR with currently-ignored pylint checks :)

File: dev/requirements.txt [hunk 5]
```diff
@@ -1,3 +1,7 @@
 jira==1.0.3
+numpy==1.11.1
+pep8==1.7.0
 PyGithub==1.26.0
+pylint==1.6.4
+Sphinx==1.4.6
 Unidecode==0.04.19
```
Comment (holdenk) [inline dev/requirements.txt]: Could you add `requests==2.11.1`? The version of pip being installed needs a new versions of the requests library that might not be in the default virtualenv everyone has on their machine (and sadly pip doesn't explicitly depend on this).

Comment (Stibbons) [inline dev/requirements.txt]: done

**Comment by holdenk** [comment]
So looking at https://github.com/ContinuumIO/anaconda-issues/issues/305 it seems like we might cause problems for Conda users (and I've setup Conda locally to verify that this is indeed the case with the current version of Conda). This is sort of a Conda bug, but it might make sense to include the workaround for Conda users in the PR (I'll make a PR against your branch with a workaround for Conda users you can take a look at @Stibbons :)).

**Comment by holdenk** [comment]
Another minor question is running the lint script now produces a lot of output about the package versions - it might make sense to silence this output except on failures (as we do with the actually linting its self). Although I can see the argument go the other way too (if the user is on a slow network connection the first time the run lint-python we need to be sure they understand its going to take awhile to download stuff).

**Comment by sethah** [comment]
I tested this on macosx, and got some pylint errors:

````
+ echo 'Checking Pylint...'
Checking Pylint...
+ for to_be_checked in '"$PATHS_TO_CHECK"'
+ '[' false == true ']'
+ pylint --rcfile=/Users/sethhendrickson/Development/spark/python/pylintrc ./python/pyspark/ ./examples/src/main/python/ ./dev/sparktestsupport ./dev/run-tests.py ./python/run-tests.py ./dev/run-tests-jenkins.py
+ '[' 6 -ne 0 ']'
+ lint_status=1
+ echo 'Pylint checks failed.'
Pylint checks failed.
+ cat /Users/sethhendrickson/Development/spark/dev/pylint-report.txt
************* Module pyspark.daemon
E:108,19: Value 'ex' is unsubscriptable (unsubscriptable-object)
************* Module pyspark.sql.streaming
E: 23,20: Using variable 'long' before assignment (used-before-assignment)
************* Module pyspark.sql.tests
W:1432, 8: Using deprecated method assertRaisesRegexp() (deprecated-method)
W:1435, 8: Using deprecated method assertRaisesRegexp() (deprecated-method)
...
````

**Comment by Stibbons** [comment]
Indeed, I have not taken into account Anaconda environment, first because this tool provide quick and efficient way of having an almost good working environment to run jobs with numpy, pandas, and so on. There are many conda packages available to download, it could almost make pip irrelevant.

But actually not. Indeed, Anaconda and pip do not really play in the same garden. Anaconda is great for one shot job or numpy script that needs to run is a quite well controlled environment, but this level of control is not that great. You can say you want numpy v1.1.0 but the user could have chosen another "channel" that the default one because someelse has compiled for the architecture. For example you can use `conda install -c intel` to download intel optimized libraries. This is great but there is too much latitude errors to occur. Reproducibility is not as much guaranteed than with pip

At least on pypi you have the default libraries and you are guaranteed to have the original package, unpatched, and if the developer has provided "wheels", you might even download the precompiled binaries (see [numpy](https://pypi.python.org/pypi/numpy) packages).

And you want to have a great control on the deployment of your master environment. Services such as heroku or similar does not provide a conda environment by default for the deployment of the application, if user wants it is can download it of course.

**TL;DR**, executing the Spark master inside a conda environment is a BAD idea. Executing Driver and Executor inside a Conda environment makes totally sense on the other hand, and this is the proposal for this PR we are working on with Jeff: #14180.

So, for this PR, for this PR we have several choices:
- trust the user: he knows what he does. If we are inside a conda environment, we can install packages with `conda install`. I must admit this is not my favorite choice, since altering the conda environment might have unwanted behavior (if you dont use the right channel, you might replace some dependencies without the user acknowledge). Using pip from within a conda environment is not recommend at all
- detect if we are inside a conda environment, leave it unconditionally inside at the script. Of course this only impact the `lint-python` script. I prefer this solution since we fallback into a nice controlled virtualenv for lint script

I use quite a lot conda nowadays, playing with sckitlearn and alike, I must admit I am well impressed by the quality of the features and precompiled packages it provides, and I find it really useful for computation jobs, even for Spark job.

For the output, I'll make better use of the `--debug` argument

**Comment by Stibbons** [comment]
I have just tested this pull request on MacOS X with latest version of Spark, no error:

```
...
Checking Pep8...
PEP8 checks passed.
Checking Pylint...
Pylint checks passed.
rm -rf _build/*
pydoc checks passed.
```

Is spark compiled with MacOS on Jenkins ?

**Comment by Stibbons** [comment]
Here is my proposal: I leave the environment and let the script create the virtual env, plus minor improvement on the linter script.

**Comment by holdenk** [comment]
@Stibbons the jenkins environment for Spark is not OS X, but since a lot of the developers work in OS X I figured it would be good to test there too. I think we should probably figure out why @sethah is getting a different result than you and I are.

As for Conda I'm not sure I 100% agree, looking at https://www.continuum.io/blog/developer-blog/using-anaconda-pyspark-distributed-language-processing-hadoop-cluster and http://blog.cloudera.com/blog/2016/02/making-python-on-apache-hadoop-easier-with-anaconda-and-cdh/ it seems like many people are looking at ways to use conda together with PySpark so I think we should consider the possibility of running on a machine with Conda installed (but we can investigate other work arounds besides the one I proposed in the PR to your branch) :)

**Comment by Stibbons** [comment]
You and I agree, actually:
- PySpark can run inside Anaconda, and indeed this is greatly valuable. This will make available to the "driver" all the package provided by Anaconda (in client mode)
- Making the same environment available for executor is a bit messy for the moment. Here we have an NFS share that every executor use, but having it automatically deployed would be so cool. that's why the other PR :)

For development, PySpark will still use the anaconda environment if the user wishes so. Just for the `lint-python` execution, we revert back to a virtualenv, but only for the duration of the script

**Comment by Stibbons** [comment]
I really think using Spark with Anaconda is a **must have**. Deploying jobs that runs inside a Conda environment is so fast and efficient. I really want to push for this pull request #14180 that also add anaconda support for executors (thanks to jeff's work). Actually before this PR I never tried Anaconda, and now I use it everyday on jupyter. Anaconda on the driver and executor make totally sense!

**Comment by Stibbons** [comment]
Ho, I have these pylint errors on my ubuntu! Probably I did not rebased correctly.

Fixed with new check ignore:
- deprecated-method
- unsubscriptable-object,
- used-before-assignment,

of course, once merged. i can work on fixing this pylint errors !

**Comment by holdenk** [comment]
@Stibbons however the problem is the virtualenv that gets created when a user is in a conda enviroment is partially broken.

**Comment by holdenk** [comment]
Also if it would be useful to have a quick chat off-line and then circle back to the PR with the result of our chat let me know - I'd really like to see us get something like this in so we can have better/more consistent linting.

**Comment by Stibbons** [comment]
I agree, just email me :)

**Comment by Stibbons** [comment]
Any hope this patch might be integrated ?

**Comment by holdenk** [comment]
That's a reasonable question, I'll try and look at this some next week but if you've got a chance to look at/address the current comments on it that would be great.

**Comment by holdenk** [comment]
So under the current jenkins set up this isn't running the Python style checks in Jenkins so it would be good to make sure these work in jenkins (can you update a python file)?

**Comment by ueshin** [comment]
Hi, are you still working on this?
